### ISSUE TITLE
PemReader.LoadCertificate should take into account certificate chains

### ISSUE THREAD

**Issue by heaths** [Azure/azure-sdk-for-net#19043]
Despite the old code in Azure.Identity's `ClientCertificateCredential` not handling certificate chains, the new `PemReader` should be able to handle it more easily as it loops through 0 or more PEM fields. We could store each found `CERTIFICATE` in a chain and, when finished, build it to figure out which is the leaf.

**Comment by heaths** [comment]
I'm starting to look into this more, and I don't think the order can be so random. Though I've not been able to find a specific RFC, https://www.rfc-editor.org/rfc/rfc2459#section-3.2 does indicate the leaf should be first followed by zero or more signing CAs. In testing this, I found that order is important with `X509Certificate2.CreateFromPemFile` such that:

1. Optional private key of leaf
2. Certificate of leaf
3. Zero or more intermediate / root CAs in any order

Even if I co-locate the private key and associated certificate anywhere in a certificate chain PEM, `CreateFromPemFile` throws a rather nondescript error,

> The key contents do not contain a PEM, the content is malformed, or the key does not match the certificate.

/cc @bartonjs for confirmation.

That said, there's still work to be done and value to be had to support chains. If a certificate is missing an intermediate certificate - even if the root is installed / trusted on the machine - the certificate cannot be properly verified.

**Comment by heaths** [comment]
After reviewing related RFCs and discussing this with the .NET cryptography team, this doesn't appear to be a common case and indicative of a malformed PEM. The PKCS8 private key and associated certificate should be first in the content, with RFC 2549 saying it should be first; though, in trial, I found `X509Certificate2.LoadFromPemFile` accepts it either as the first or second entry, with the certificate being the second or first entry.

I will open a separate feature request to consider returning a chain, but keep in mind this method is just a convenience that anyone can do themselves and, for example, handle certificate chains themselves. See https://learn.microsoft.com/samples/azure/azure-sdk-for-net/get-certificate-private-key/ for the original sample I wrote to cover the scenario that `DownloadCertificate` does now.

**Comment by christothes** [comment]
We have a customer issue where the cert chain issue came up https://github.com/Azure/azure-sdk-for-net/issues/30318

**Comment by heaths** [comment]
We'll have to write it ourselves then. .NET does not support it.

**Comment by heaths** [comment]
[`X509Certificate2.CreateFromPemFile`](https://learn.microsoft.com/dotnet/api/system.security.cryptography.x509certificates.x509certificate2.createfrompemfile) was added in net5.0, and since we support netstandard2.0 we'll need to write it. Maybe we can use either reflection or multi-targeting (right now, KV doesn't multi-target because of the permutations required and we'd still have to reflect around netstandard2.0's "lies" about `System.Security.Cryptography` anyway), but I'd rather us be consistent in our implementation so it's easier to diagnose and maintain expectations.

I did some prototyping before and it's straight forward to support common formats. The investigation above was about where the `PRIVATE KEY` (PKCS7) or `RSA PRIVATE KEY` (RSA1_5) could reasonably go in the certificate chain, which .NET assumes as well.

**Comment by pallavit** [comment]
I agree with your over all stance on support should be consistent and lighting this feature only for .NET5.0 and beyond may not be correct.  

However, does this mean today we can't support certificate chains at all? If so, may be it is pragmatic to provide the support for .NET5.0 and beyond and if we do see multiple customer requests for older versions we will consider adding support for them too as needed. Also, I am assuming this means we will need to multitarget KV. What do you think?

**Comment by heaths** [comment]
I agree with supporting this only when .NET supports it, though as @KrzysztofCwalina pointed out previously, the public APIs we expose need to exist for all TFMs. They can throw, though, a `NotSupportedException`. We don't necessarily need to multi-target as long as we use the trimming-compatible reflection light-up, but it may be easier to multi-target in the long run.

However, in that case, we may run afoul of previously stated goals from one of our last team meetings where @jsquire had probably the most to say about this: do we bifurcate on net5.0 that introduced the API we want to use, or net6.0 since that's the min supported .NET currently?

**Comment by heaths** [comment]
That said, it's not exactly hard to support ourselves. I previously threw together an admittedly kinda of gross sample - just a quick and dirty proof of concept we can build on: https://gist.github.com/heaths/d71d646468d5852ec712a9d6e2ef0816

**Comment by jsquire** [comment]
> However, in that case, we may run afoul of previously stated goals from one of our last team meetings where @jsquire had probably the most to say about this: do we bifurcate on net5.0 that introduced the API we want to use, or net6.0 since that's the min supported .NET currently?

We'd go with `net6.0` if we're introducing the target now.  `net5.0` is an unsupported runtime.

**Comment by heaths** [comment]
> We'd go with `net6.0` if we're introducing the target now. `net5.0` is an unsupported runtime.

In that case, I think the method should throw `NotSupportedException` downlevel then, at least if we find multiple PEM blocks. If we don't actually support loading a certificate chain we shouldn't accept one as input at all. The difference in behavior would be jarring.

**Comment by christothes** [comment]
throwing for down-level seems reasonable. At least we'd present a better error message than today, which, I believe, silently provides the first cert in the chain.
